### PR TITLE
Speech Recognition

### PR DESCRIPTION
**Pull request by tl676**
Implemented Speech Recognition from front-end to back-end, including new StoppableThread class to deal with concurrency issues.  Also has the implementation for the front-end button + message box.

### CODE REVIEW

File: basestation/base_station_interface.py [hunk 1]
```diff
@@ -289,6 +297,122 @@ def get(self):
...
+    def post(self):
+        """ Starts and stops the speech recognition thread depending on the 
+        command sent in the request  
+        """
+        data = json.loads(self.request.body.decode())
+        # The command is either START or STOP
+        command = data["command"]
+        bot_name = data["bot_name"]
+        bot_id = self.base_station.bot_name_to_bot_id(bot_name)
+        session_id = self.get_secure_cookie("user_id")
```
Comment (virenvshah) [inline basestation/base_station_interface.py:330]: Add comment to either remove or deal with session ids.

**Review by slawrence100 — CHANGES_REQUESTED** [review]
I had a few installation issues here - TL;DR we'll need to update our installation files / docs first. Here's what happened:

1. I tried running the base station, but it told me that there's no package named `pyaudio`. We should probably add this to the requirements file.
2. I tried `pip install pyaudio`, but that produces this error:
```
src/_portaudiomodule.c:29:10: fatal error: 'portaudio.h' file not found
    #include "portaudio.h"
             ^~~~~~~~~~~~~
    1 error generated.
    error: command 'gcc' failed with exit status 1
```
3. I ended up installing `wheel` before this upon getting a similar error. 
4. I found a solution - [StackOverflow](https://stackoverflow.com/questions/33513522/when-installing-pyaudio-pip-cannot-find-portaudio-h-in-usr-local-include) said to install `portAudio` via homebrew.
5. The `pip install pyaudio` worked fine afterwards

Finally, I don't think the system recognized anything I said - it continuiously said "say something" followed by a timeout even though my system settings say my mic is fine. I noticed that neither Firefox nor Chrome prompted me for mic permissions - maybe that's part of the issue?

edit: The issue with recognition was running the base station in a VSCode terminal, which doesn't allow one to get hardware (i.e. mic) permissions. Once I ran it in Terminal, everything works as expected.

**Review by slawrence100 — APPROVED** [review]
Looks good - Thanks for the installation updates!

**Review by kabirwalia8300 — APPROVED** [review]
Looks good!

**Review by ashok-arun — APPROVED** [review]

**Pull request merged**